cmd/libsnap-confine-private/mount-opt.c

```diff
@@ -115,3 +118,134 @@ const char *sc_mount_opt2str(unsigned long flags)
+	// - bind mounts (bind)
+	//   - including the recursive variant
+	if (mountflags & MS_BIND) {
+		const char *special = mountflags & MS_REC ?
+		    " --rbind" : " --bind";
+		used_special_flags |= MS_BIND | MS_REC;
+		to = stpcpy(to, special);
+	}
+	// - moving mount point location (move)
+	if (mountflags & MS_MOVE) {
+		const char *special = " --move";
+		used_special_flags |= MS_MOVE;
+		to = stpcpy(to, special);
+	}
+	// - shared subtree operations (shared, slave, private, unbindable)
+	//   - including the recursive variants
+	if (MS_SHARED & mountflags) {
+		const char *special = mountflags & MS_REC ?
+		    " --make-rshared" : " --make-shared";
+		used_special_flags |= MS_SHARED | MS_REC;
+		to = stpcpy(to, special);
+	}
+	if (MS_SLAVE & mountflags) {
+		const char *special = mountflags & MS_REC ?
+		    " --make-rslave" : " --make-slave";
+		used_special_flags |= MS_SLAVE | MS_REC;
+		to = stpcpy(to, special);
+	}
+	if (MS_PRIVATE & mountflags) {
+		const char *special = mountflags & MS_REC ?
+		    " --make-rprivate" : " --make-private";
+		used_special_flags |= MS_PRIVATE | MS_REC;
+		to = stpcpy(to, special);
+	}
+	if (MS_UNBINDABLE & mountflags) {
+		const char *special = mountflags & MS_REC ?
+		    " --make-runbindable" : " --make-unbindable";
+		used_special_flags |= MS_UNBINDABLE | MS_REC;
+		to = stpcpy(to, special);
+	}
+	// If regular option syntax exists then use it.
+	if (mountflags & ~used_special_flags) {
+		const char *regular =
+		    sc_mount_opt2str(mountflags & ~used_special_flags);
+		to = stpcpy(to, " -o ");
+		to = stpcpy(to, regular);
+	}
+	// Add source and target locations
+	if (source != NULL && strcmp(source, "none") != 0) {
+		to = stpcpy(to, " ");
+		to = stpcpy(to, source);
+	}
+	if (target != NULL && strcmp(target, "none") != 0) {
+		to = stpcpy(to, " ");
+		to = stpcpy(to, target);
+	}
+	return buf;
+}
+
+char *sc_umount_cmd(const char *target, int flags)
+{
+	// NOTE: this uses static buffer because it has lower complexity than a
+	// dynamically allocated buffer. We've decided as a team to prefer this
+	// approach.
+	static char buf[PATH_MAX + 1000];
+	char *to = buf;
+
+	to = stpcpy(to, "umount");
+
+	if (flags & MNT_FORCE) {
+		to = stpcpy(to, " --force");
+	}
+
+	if (flags & MNT_DETACH) {
+		to = stpcpy(to, " --lazy");
+	}
+	if (flags & MNT_EXPIRE) {
+		// NOTE: there's no real command line option for MNT_EXPIRE
+		to = stpcpy(to, " --expire");
+	}
+	if (flags & UMOUNT_NOFOLLOW) {
+		// NOTE: there's no real command line option for UMOUNT_NOFOLLOW
+		to = stpcpy(to, " --no-follow");
+	}
+	if (target != NULL) {
+		to = stpcpy(to, " ");
+		to = stpcpy(to, target);
+	}
+	return buf;
+}
+
+void sc_do_mount(const char *source, const char *target,
+		 const char *filesystemtype, unsigned long mountflags,
+		 const void *data)
+{
+	char *mount_cmd =
+	    sc_mount_cmd(source, target, filesystemtype, mountflags, data);
+	debug("performing operation: %s", mount_cmd);
```

Comment:
I don't care for how we have sc_mount_cmd() and sc_umount_cmd() unconditionally are doing all kinds of string operations only for debug output that may not be displayed. IMO, it would be better to create a separate debug_mount() and debug_umount() and have them exit if SNAP_CONFINE_DEBUG is not set and only conditionally compiled with build flags (same for debug() for that matter).